File: python_solution/problem19.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Created on Sun Sep  8 21:09:10 2024

@author: sharminalina
"""
from typing import Optional

class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Problem19:
    
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        dummy = ListNode(0);
        dummy.next = head;

        fast = dummy;
        slow = dummy;
        for i in range(n+1):
            fast = fast.next
        
        while fast is not None:
            fast = fast.next
            slow = slow.next

            
            
        slow.next = slow.next.next;

        return dummy.next;
        
    
    def CreateLinkedList(self, nums: list[int])->ListNode:
        if len(nums) == 0:
            return None

        head = ListNode(nums[0])
        current = head
        for i in range(1,len(nums)):
            current.next = ListNode(nums[i])
            current = current.next
        return head
    
    def printList(self, head: ListNode) -> None:
        current = head
        while current is not None:
            print(current.val, end=" ")
            current = current.next
        print()

            



def main():
    print("Problem No. 19 : Remove Nth node from End of list")
    
    nums_input = input("Enter the list of integers (comma-separated): ")
    # Convert input string to a list of integers
    nums = list(map(int, nums_input.split(',')))
    
    
    solution_prob19 = Problem19()
    
    head = solution_prob19.CreateLinkedList(nums)
    
    solution_prob19.printList(head)
    
    n = int(input("Enter the value of n (nth node from the end to be removed): "))
    
    # Call RemoveNthFromEnd to remove the node
    
    result = solution_prob19.removeNthFromEnd(head, n);

    # Print the modified list
    print("Modified List:");
    solution_prob19.printList(result)

if __name__ == "__main__":
    main()

File: C#Solutions/problem19.cs
using System;

namespace AllProblems
{
    

    public static class Problem19
    {
        public class ListNode
        {
            public int val;
            public ListNode next;
            public ListNode(int val = 0, ListNode next = null)
            {
            this.val = val;
            this.next = next;
            }
        }
        
        // This method removes the nth node from the end of the list
        public static ListNode RemoveNthFromEnd(ListNode head, int n)
        {
            ListNode dummy = new ListNode(0);
            dummy.next = head;

            ListNode fast = dummy;
            ListNode slow = dummy;

            // Move the fast pointer n+1 steps forward
            for (int i = 0; i <= n; i++)
            {
                fast = fast.next;
            }

            // Move both fast and slow until fast reaches the end
            while (fast != null)
            {
                fast = fast.next;
                slow = slow.next;
            }

            // Remove the nth node
            slow.next = slow.next.next;

            return dummy.next;
        }

        // This method solves the problem and is called from Program.cs
        public static void Solve()
        {
            Console.WriteLine("Enter the values for the linked list (comma-separated): ");
            string[] input = Console.ReadLine().Split(',');

            // Convert the input string to integers and create the linked list
            ListNode head = CreateLinkedList(input);

            // Print the original list
            Console.WriteLine("Original List:");
            PrintList(head);

            // Read input for n (nth node from the end)
            Console.Write("\nEnter the value of n (nth node from the end to be removed): ");
            int n = int.Parse(Console.ReadLine());

            // Call RemoveNthFromEnd to remove the node
            ListNode result = RemoveNthFromEnd(head, n);

            // Print the modified list
            Console.WriteLine("\nModified List:");
            PrintList(result);
        }

        public static ListNode CreateLinkedList(string[] input)
        {
            if (input.Length == 0)
                return null;

            ListNode head = new ListNode(int.Parse(input[0]));
            ListNode current = head;

            for (int i = 1; i < input.Length; i++)
            {
                current.next = new ListNode(int.Parse(input[i]));
                current = current.next;
            }

            return head;
        }

        // Helper method to print the list
        public static void PrintList(ListNode head)
        {
            ListNode current = head;
            while (current != null)
            {
                Console.Write(current.val + " ");
                current = current.next;
            }
            Console.WriteLine();
        }
    }
}

        

       
        
            
        

        
